llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp

```diff
@@ -3224,8 +3224,25 @@ bool RISCVDAGToDAGISel::selectInvLogicImm(SDValue N, SDValue &Val) {
 
   // Abandon this transform if the constant is needed elsewhere.
   for (const SDNode *U : N->users()) {
-    if (!ISD::isBitwiseLogicOp(U->getOpcode()))
+    switch (U->getOpcode()) {
+    case ISD::AND:
+    case ISD::OR:
+    case ISD::XOR:
+      if (!(Subtarget->hasStdExtZbb() || Subtarget->hasStdExtZbkb()))
+        return false;
+      break;
+    case RISCVISD::VMV_V_X_VL:
+      if (!Subtarget->hasStdExtZvkb())
+        return false;
+      if (!all_of(U->users(), [](const SDNode *V) {
+            return V->getOpcode() == ISD::AND ||
+                   V->getOpcode() == RISCVISD::AND_VL;
```

Comment:
Shall I check if that's not the passthru operand? If so, how to test it?

Reply:
ISD::AND doesn't have a passthru. For RISCV::AND_VL you would need to loop through `U->uses()` instead of `U->users()`. That will give you an `SDUse &`. From there you can call `SDUse::getUser()` to get `V` and `SDUse::getOperandNo` to get the operand number of the use. The passthru will be operand 2.

Reply:
How to test? Is `@llvm.riscv.vand`'s first operand the passthru?

Would it be possible for the same `VMV_V_X_VL` to be the AND_VL passthru and one of the AND_VL "and" operands at the same time?

Would the following work:

```cpp
      if (!all_of(U->users(), [U](const SDNode *V) {
             return V->getOpcode() == ISD::AND ||
                   (V->getOpcode() == RISCVISD::AND_VL &&
                    U != V->getOperand(2).getNode()); // not passthru
           }))
```

?

Reply:
Will you approve this PR as-is then?
I don't think adding untested code is a good idea. In the unlikely event of `U` somehow appearing in passthru, it would only lead to somewhat suboptimal code.